tensorflow-keras/src/main/java/org/tensorflow/keras/optimizers/Adamax.java

```diff
@@ -0,0 +1,283 @@
+    float epsilon = (float) config.getOrDefault(EPSILON_KEY, EPSILON_DEFAULT);
+    float betaOne = (float) config.getOrDefault(BETA_ONE_KEY, BETA_ONE_DEFAULT);
+    float betaTwo = (float) config.getOrDefault(BETA_TWO_KEY, BETA_TWO_DEFAULT);
+    if (name == null) {
+      return new Adamax(tf, learningRate, betaOne, betaTwo, epsilon);
+    } else {
+      return new Adamax(tf, name, learningRate, betaOne, betaTwo, epsilon);
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Map<String, Object> getConfig() {
+    return config;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public float getLearningRate() {
+    return this.learningRate;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void setLearningRate(float learningRate) {
+    this.learningRate = learningRate;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  protected Optional<Op> prepare(String scopeName) {
+    betaOneConst = tf.constant(betaOne);
+    betaTwoConst = tf.constant(betaTwo);
+    learningRateConst = tf.constant(learningRate);
+    epsilonConst = tf.constant(epsilon);
+
+    return Optional.empty();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  protected void createSlots(List<Output<? extends TType>> variables) {
+    for (Output<? extends TType> v : variables) {
+      createAdamaxSlot(v.asOutput());
+    }
+    betaOnePower = tf.withName("beta1_power").variable(Shape.scalar(), TFloat32.DTYPE);
+    Assign<TFloat32> betaOnePowerInit = tf.assign(betaOnePower, tf.constant(betaOne));
+    ((Graph) tf.scope().env()).addInitializer(betaOnePowerInit);
+  }
+
+  /**
+   * Create the first and second moment slots
+   *
+   * @param v the variable
+   * @param <T> the datatype of the variable
+   */
+  private <T extends TType> void createAdamaxSlot(Output<T> v) {
+    Operand<T> firstMomentInitializer =
+        tf.fill(tf.shape(v), tf.dtypes.cast(tf.constant(0.0f), v.dataType()));
+    createSlot(v.asOutput(), FIRST_MOMENT, firstMomentInitializer);
+    Operand<T> secondMomentInitializer =
+        tf.fill(tf.shape(v), tf.dtypes.cast(tf.constant(0.0f), v.dataType()));
+    createSlot(v.asOutput(), SECOND_MOMENT, secondMomentInitializer);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  protected <T extends TType> Op applyDense(Output<T> gradient, Output<T> variable) {
+    Variable<T> firstMomentSlot = getSlot(variable, FIRST_MOMENT).get();
+    Variable<T> secondMomentSlot = getSlot(variable, SECOND_MOMENT).get();
+    return ApplyAdaMax.create(
+        scope,
+        (Operand) variable,
+        (Operand) firstMomentSlot,
+        (Operand) secondMomentSlot,
+        (Operand) tf.dtypes.cast(betaOnePower, gradient.dataType()),
+        (Operand) tf.dtypes.cast(learningRateConst, gradient.dataType()),
+        (Operand) tf.dtypes.cast(betaOneConst, gradient.dataType()),
+        (Operand) tf.dtypes.cast(betaTwoConst, gradient.dataType()),
+        (Operand) tf.dtypes.cast(epsilonConst, gradient.dataType()),
+        (Operand) gradient);
```

Comment:
Again are you sure we need these explicit casting to `Operand`? If so, something is wrong with our generic signatures and we should address that.

Reply:
I removed the (Operand) casts. It must have been when I was just figuring out the framework.